segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -701,9 +711,48 @@ public void close() {
             } catch (Exception e) {
                 throw new CompletionException(e);
             }
+        }, this.executor)
+        .whenCompleteAsync((v, e) -> {
```

Comment:
This is a good example of when `Async` is not needed.

Reply:
In general any whenCompleteAsync without executor will eat up processing from core thread pool and also there is no guarantee that it won't run on fork join pool. (The documentation say this can run on fork join pool).
Storage I/O thread pool is very IO heavy - these few lines and waiting on IO is all these threads are doing.
Plus it is using the same thread pool every where so there is actually no context switching.

Reply:
http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/CompletableFuture.java#l2372 This stuff scared me.